src/expr/src/scalar/func.rs

```diff
@@ -2791,6 +2769,7 @@ impl BinaryFunc {
 
             ParseIdent => ScalarType::Array(Box::new(ScalarType::String)).nullable(in_nullable),
             PrettySql => ScalarType::String.nullable(in_nullable),
+            RegexpReplace { .. } => ScalarType::Array(Box::new(ScalarType::String)).nullable(true),
```

Comment:
Also, it shouldn't be an array. It should be exactly the same as for the original, variadic `RegexpReplace`, see [here](https://github.com/MaterializeInc/materialize/blob/f89bb6051898d7c8aa7e5cd39816921b7e8ae86d/src/expr/src/scalar/func.rs#L7871).

This is the cause of the CI test failure in `Fast SQL logic tests`. You can reproduce this locally by:
```
RUST_BACKTRACE=1 bin/sqllogictest -- -v --auto-index-selects test/sqllogictest/redacted.slt
```

Note that the reason why `regex.slt` didn't catch this is that it's not run with `--auto-index-selects` (because [it's in the](https://github.com/MaterializeInc/materialize/blob/77275f5565e001e9abaadea9b041f4c97686d646/test/sqllogictest/mzcompose.py#L516) `tests_without_views` category). If you run `regex.slt` with `--auto-index-selects` locally, then it also happens there. (What `--auto-index-selects` does is that for each SELECT in the slt file, it also creates an indexed view of it, and selects from the indexed view. This is to catch bugs that happen only with a long-running dataflow, and not for a one-off SELECT.) 

But I don't understand why is this happening only with `--auto-index-selects`. We should definitely figure this out.

Reply:
I'm also not sure why is `regex.slt` in the `tests_without_views` category. Maybe we should remove it from there. On `main`, it runs fine with `--auto-index-selects`. @MaterializeInc/testing, do you know what are the possible reasons in general why we put slts in `tests_without_views`? I thought it's only when it would fail with `--auto-index-selects` due to legitimate reasons (e.g., different EXPLAIN output for a long-running dataflow). Is it also due to slowness reasons? If `regex.slt` was there only due to slowness reasons, then I'd say running `regex.slt` with `--auto-index-selects` would be actually worth the time. There have been so many tricky regex bugs that testing them as thoroughly as possible is important.

Reply:
Unfortunately, I don't have a good understanding of why files are in a certain category. Although, I touched the test file, I only undertook semantic-preserving refactorings.
The comment in the code "renable a fast subset of the following when performance is restored" hints that performance can be a reason for a file to be assigned to `tests_without_views`.

Feel free to change assignments if you are somewhat confident with your change!

Reply:
@sdht0, could you try removing `"test/sqllogictest/regex.slt",` from the `tests_without_views` list? I think it should run fine after fixing the type bug.

Reply:
Thanks for the review @ggevay! I will address the comments later today.

Reply:
Oh, I think I know why the type bug was getting caught by slts only with `--auto-index-selects`: The check that was failing is based on the `RelationDesc`, and I think for one-off SELECTs the `RelationDesc` is created based on the HIR plan (see `plan_root_query`), which is before the MIR optimizations are run, but for an indexed view we rely on the type that is after MIR optimization. @mgree, @aalexandrov, one more reason why we should add a check that the type of a query after optimization matches the type that was before optimization.